### ISSUE TITLE
[git-server] Implement `pre-receive` git hook for verifying a signed git push to authorize an update

### ISSUE THREAD

**Issue by Ryanmtate** [radicle-dev/radicle-client-services#39]
The `git-server` currently uses http basic authorization headers (i.e. username and password) to set the `REMOTE_USER`, used by the `git-http-backend`. At the moment, `REMOTE_USER` is not being used to verify a remote sender against an authority.

There are a few ways we can go about authorizing a git request:

- Use HTTP BASIC Authorization;
- Use SSH identity;
- Use GPG signed push (i.e. `git push --signed ...`);

The first two are commonly used by centralized git services, such as `GitHub` and `GitLab`, and are therefore well known methods for authentication. 

GPG signed pushes, however, are less known and also less supported by centralized git services. However, in the context of a p2p decentralized version of git (i.e. Radicle), using GPG signed pushes may be a better solution, as [it more explicitly states the intent of updating a reference/branch](https://people.kernel.org/monsieuricon/signed-git-pushes), which basic auth and SSH, by itself, do not.*

Using GPG signed push, we can receive useful environmental variables set by the http-backend that can be used by the `pre-receive` hook to validate the request before any changes are made. If `pre-receive` hooks fails, no changes are made to the target. This is the first hook called by the `git-receive-pack` event, and is also the ideal place for authenticating a GPG certificate.

The [certificate environment variables](https://git-scm.com/docs/git-receive-pack/2.13.7#_pre_receive_hook) we receive include:

```
GIT_PUSH_CERT
GIT_PUSH_CERT_SIGNER
GIT_PUSH_CERT_KEY
GIT_PUSH_CERT_STATUS
GIT_PUSH_CERT_NONCE
GIT_PUSH_CERT_NONCE_STATUS
GIT_PUSH_CERT_NONCE_SLOP
```

`GIT_PUSH_CERT` provides the object name of the blob where the signed push certificate exists.

`GIT_PUSH_CERT_SIGNER` provides the name and email address of the signer, e.g. `Ryan Tate <ryan.michael.tate@gmail.com>`.

`GIT_PUSH_CERT_KEY` provides the GPG key ID used to sign the certificate. This can be found using the following command, `gpg --list-secret-keys --keyid-format=long`, which produces something like:

```
/Users/ryantate/.gnupg/pubring.kbx
----------------------------------
sec   rsa2048/817EEFE32E1F0AA5 2020-04-03 [SC] [expires: 2022-04-03]
      A26B6C38CF03E732B95E81A4817EEFE32E1F0AA5
uid                 [ultimate] Ryan Tate <ryan.michael.tate@gmail.com>
ssb   rsa2048/19421653C2A049D9 2020-04-03 [E] [expires: 2022-04-03]
```

In this case, the matching `GIT_PUSH_CERT_KEY` for the GPG key above would be `817EEFE32E1F0AA5`.

`GIT_PUSH_CERT_STATUS` provides the status on the verified GPG signed push certificate. A status of `OK` means the certificate is valid, and contains the correct challenge nonce requested by the git-server. If the status is `UNSOLICITED`, `MISSING`, `BAD` or `SLOP`, then the certificate has failed verification and we can decline the request.

For authentication purposes, a valid certificate is only the first step for authorization. The `GIT_PUSH_CERT_KEY` can be used to check the authority of the signer for the git-server or namespace repo.

- One approach could be setting an `AUTHORIZED_KEYS` environmental variable that includes a comma delimited string of GPG key IDs, e.g. `817EEFE32E1F0AA5,...,...` or a path to a file containing the key IDs. 

- Another approach is to use ENS to store GPG key IDs for lookup, mapping to an organization. 

- Yet another approach would be to keep a file in the namespace repo called `.authorized_keys` that includes the GPG key IDs, and whenever a peer would like to push an update, that file is referenced on the target branch for whether the sender may push an update.

----
Footnotes:

*This issue is not suggesting abandoning basic auth and SSH authentication, but providing those features in tandem with signed git pushes. basic auth, for example, could be feature gated, and used for other authentication purposes. However, it may be deemed that basic auth is made extraneous by use of signed git pushes. TBD.

**Comment by Ryanmtate** [comment]
One thought on using a `.authorized_keys` file, it needs to be able to specify branch based authentication. Need to think through the file format for this, but should be humanly readable and editable, e.g. JSON, TOML, YAML, etc.

**Comment by cloudhead** [comment]
Nice write up, the approach is promising!

In terms of branch-based permissions, I'm not sure it's going to make sense for users to push to the same repository, in which case the auth would not need to be per-branch.

Contributors will have their own fork anyway, as this is how Radicle works: it's a bazaar model. I think the problem of deciding on some kind of canonical master should be decoupled from the push flow. Sharing a single repository amongst contributors is not going to map very well to the peer to peer world.

**Comment by Ryanmtate** [comment]
> In terms of branch-based permissions, I'm not sure it's going to make sense for users to push to the same repository, in which case the auth would not need to be per-branch.

I don't think that is the suggestion here, I think the problem is guarding against a pushing to a branch that they do not own.

> In our case I'd expect that all maintainers/members of an org are allowed to push their code to the org node, under their own namespace/tree.

What guarantees this?

**Comment by Ryanmtate** [comment]
Closed by #38
